### ISSUE TITLE
False non-exhaustive match warning for float(x) class pattern

### ISSUE THREAD

**Issue by mbergin** [microsoft/pyright#10837]
## Environment data

- Pylance version: 2025.7.1
- OS and version: Linux desktop 5.15.167.4-microsoft-standard-WSL2
- Python version (& distribution if applicable, e.g. Anaconda): CPython 3.13.5 [Clang 20.1.4] (main, Jul 11 2025, 22:43:46)

## Code Snippet

```python
match 1.0:
    case float(x):
        pass
```

## Repro Steps

- Set `"python.analysis.typeCheckingMode": "strict"` in settings.json.
- Save the snippet and open it in VS Code with Pylance enabled.
- Observe the diagnostic on the `match` statement.

## Expected behavior

- No exhaustiveness diagnostic.  

## Actual behavior

```
Cases within match statement do not exhaustively handle all values
  Unhandled type: "int"
  If exhaustive handling is not intended, add "case _: pass"Pylance[reportMatchNotExhaustive](https://github.com/microsoft/pylance-release/blob/main/docs/diagnostics/reportMatchNotExhaustive.md)
```

**Comment by LeeeeT** [comment]
In Python's typing community `int` is considered a subtype of `float` (contradicting the run time). Basically, every time you see `float` in an annotation it is actually `float | int`. See the following example:
```python
def f(x: float):
    match x:
        case float():
            print("float")
        case int():
            print("int")

f(1.0)  # prints "float"
f(1)  # prints "int"
```
Type checkers should not mark the last line as an error. Personally, I don't like this, but this is an established convention that is not going to change.

Pyright (or rather Python's type system) doesn't have a concept of "strict" `float` that can't be an `int`, so `1.0` is inferred simply as `float`, thus requiring you to handle both `float(x)` and `int(x)` cases.

**Comment by mbergin** [comment]
Thanks. Now you've told me this I've managed to find some references that I'll leave here for future confused souls:

- [Stackoverflow: Why does mypy consider "int" a subtype of "float"?](https://stackoverflow.com/questions/59619778/why-does-mypy-consider-int-a-subtype-of-float)
- [mypy: Duck type compatibility](https://mypy.readthedocs.io/en/latest/duck_type_compatibility.html)
